# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Get protein sequences for a vector of uniprot IDs
#' 
#' @title Get protein IDs in fasta file for a vector of Uniprot IDs
#' @param fastaPath path to fasta formated file to look up protein sequences
#' @param ids CharacterVector of uniprot IDs
#' @return CharacterVector of protein sequences in same order as ids.
#'
#' 
getSquences <- function(fastaPath, ids) {
    .Call(`_peptideUtils_getSquences`, fastaPath, ids)
}

#' Get locations of modified residues in parent protein
#'
#' @title Get locations of modified residues in parent protein
#' @param fastaPath path to fasta formated file to look up protein sequences
#' @param ids CharacterVector of Uniprot IDs
#' @param peptideSeq CharacterVector of peptide sequences containing modifications
#' @param modSep delimiter for multiple modifications
#' @return CharacterVector containing locations of modifications in protein sequence
#'
#' 
getModifiedResidues <- function(fastaPath, ids, peptideSeq, modSep = "|") {
    .Call(`_peptideUtils_getModifiedResidues`, fastaPath, ids, peptideSeq, modSep)
}

#' Combined concated mods from multiple peptides into a single string
#' 
#' @title Combined mods from multiple peptides into a single string
#' @param mods Modifications to combine
#' @param sep delimiter separating modifications
#' @return Modifications combined into a single string
#'
#' 
combineMods <- function(mods, sep = '|') {
    .Call(`_peptideUtils_combineMods`, mods, sep)
}

#' Calculate peptide monoisotopic or average masses.
#' 
#' @title Calculate peptide masses
#' @param sequences Peptide sequences
#' @param monoMass Should monoisotopic mass be calculated. If false, average mass is calculated.
#' @param residueAtoms Path to residueAtoms file. If blank, the default file included in the package is used. 
#' @param atomMasses Path to atomMasses file. If blank, the default file included in the package is used.
#' @return vector of peptide masses.
#'
calcMass <- function(sequences, monoMass = TRUE, residueAtoms = "", atomMasses = "") {
    .Call(`_peptideUtils_calcMass`, sequences, monoMass, residueAtoms, atomMasses)
}

#' Calculate peptide molecular formulas
#' 
#' @title Calculate peptide formulas
#' @param sequences Peptide sequences
#' @param subscripts Should formulas have subscripts or normal baseline numbers?
#' @param residueAtoms Path to residueAtoms file. If blank, the default file included in the package is used. 
#' @return vector of peptide formulas.
#'
calcFormula <- function(sequences, subscripts = FALSE, residueAtoms = "") {
    .Call(`_peptideUtils_calcFormula`, sequences, subscripts, residueAtoms)
}

#' Convert from 1 letter amino acid codes to 3
#' 
#' @title Convert to 3 letter amino acid codes
#' @param sequences vector of sequences
#' @param sep_in deliminator between amino acids in input
#' @param sep_out deliminator between amino acids in output
#' @param n_term_out string to append to n terminus
#' @param c_term_out string to append to c terminus
#' @return StringVector of peptides with three letter amino acid codes
#' 
oneLetterToThree <- function(sequences, sep_in = "", sep_out = "", n_term_out = "", c_term_out = "") {
    .Call(`_peptideUtils_oneLetterToThree`, sequences, sep_in, sep_out, n_term_out, c_term_out)
}

