# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Get protein sequences for a vector of uniprot IDs
#' 
#' @title Get protein IDs in fasta file for a vector of Uniprot IDs
#' @param ids CharacterVector of uniprot IDs
#' @param fastaPath path to fasta formated file to look up protein sequences. 
#' @return CharacterVector of protein sequences in same order as ids.
#' 
#' @examples
#' #By default the fasta file included in the package containing human protein sequences is used.
#' getSequences(c("A0MZ66", "A6NMY6", "O00213", "O00213"))
#' 
#' #A fasta file can also be manually spedified.
#' fasta_path <- system.file('extdata/Human_uniprot-reviewed_20171020.fasta', package = 'peptideUtils')
#' getSequences(c("A0MZ66", "A6NMY6", "O00213", "O00213"), fasta_path)
#' 
getSequences <- function(ids, fastaPath = "") {
    .Call(`_peptideUtils_getSequences`, ids, fastaPath)
}

#' Get n residues before query in ref. If n overruns ref, the maximum number of characters will be returned.
#' 
#' @title get n residues before query.
#' @param query String to search for.
#' @param ref String to search in.
#' @param n Number of residues in output.
#' @param noExcept Should an std::runtime_error be thrown if query is not in ref?
#' 
#' @return n residues before query.
#'
nBefore <- function(query, ref, n = 1L, noExcept = FALSE) {
    .Call(`_peptideUtils_nBefore`, query, ref, n, noExcept)
}

#' Get n residues after query in ref. If n overruns ref, the maximum number of characters will be returned.
#' 
#' @title get n residues after query.
#' @param query String to search for.
#' @param ref String to search in.
#' @param n Number of residues in output.
#' @param noExcept Should an std::runtime_error be thrown if query is not in ref?
#' 
#' @return n residues after query.
#'
nAfter <- function(query, ref, n = 1L, noExcept = FALSE) {
    .Call(`_peptideUtils_nAfter`, query, ref, n, noExcept)
}

#' Get locations of modified residues in parent protein
#'
#' @title Get locations of modified residues in parent protein
#' @param fastaPath path to fasta formated file to look up protein sequences
#' @param ids CharacterVector of Uniprot IDs
#' @param peptideSeq CharacterVector of peptide sequences containing modifications
#' @param modSep delimiter for multiple modifications
#' @return CharacterVector containing locations of modifications in protein sequence
#'
#' @examples
#' getModifiedResidues(c("Q00839", "Q9HCS7", "Q7L014"), c("APQC*LGK", "FADMEC*K", "GAEIIVC*TPGR"))
#' 
getModifiedResidues <- function(ids, peptideSeq, fastaPath = "", modSep = "|") {
    .Call(`_peptideUtils_getModifiedResidues`, ids, peptideSeq, fastaPath, modSep)
}

#' Combined concated mods from multiple peptides into a single string.
#' 
#' @title Combined mods from multiple peptides into a single string
#' @param mods Modifications to combine
#' @param sep delimiter separating modifications
#' @return Modifications combined into a single string
#' 
#' @examples
#' combineMods(c('C157', 'C157|C125', 'C50', 'C125'))
#' 
combineMods <- function(mods, sep = '|') {
    .Call(`_peptideUtils_combineMods`, mods, sep)
}

#' Calculate peptide monoisotopic or average masses.
#' 
#' @title Calculate peptide masses
#' @param sequences Peptide sequences
#' @param monoMass Should monoisotopic mass be calculated. If false, average mass is calculated.
#' @param residueAtoms Path to residueAtoms file. If blank, the default file included in the package is used. 
#' @param atomMasses Path to atomMasses file. If blank, the default file included in the package is used.
#' @return vector of peptide masses.
#'
#' @examples
#' calcMass(c("ACLLPETVNMEEYPYDAEY", "ALCAEFK"))
#'
calcMass <- function(sequences, monoMass = TRUE, residueAtoms = "", atomMasses = "") {
    .Call(`_peptideUtils_calcMass`, sequences, monoMass, residueAtoms, atomMasses)
}

#' Calculate peptide molecular formulas
#' 
#' @title Calculate peptide formulas
#' @param sequences Peptide sequences
#' @param subscripts Should formulas have subscripts or normal baseline numbers?
#' @param residueAtoms Path to residueAtoms file. If blank, the default file included in the package is used. 
#' @return vector of peptide formulas.
#'
#' @examples
#' calcFormula(c("ACLLPETVNMEEYPYDAEY", "ALCAEFK"), subscripts = TRUE)
#'
calcFormula <- function(sequences, subscripts = FALSE, residueAtoms = "") {
    .Call(`_peptideUtils_calcFormula`, sequences, subscripts, residueAtoms)
}

#' Convert from 1 letter amino acid codes to 3
#' 
#' @title Convert to 3 letter amino acid codes
#' @param sequences vector of sequences
#' @param sep_in deliminator between amino acids in input
#' @param sep_out deliminator between amino acids in output
#' @param n_term_out string to append to n terminus
#' @param c_term_out string to append to c terminus
#' @return StringVector of peptides with three letter amino acid codes
#' 
#' @examples
#' oneLetterToThree(c("AC*LLPETVNMEEYPYDAEY", "ALCAEFK", "AQUPIVER", "C*TGGEVGATSALAPK"))
#' 
oneLetterToThree <- function(sequences, sep_in = "", sep_out = "", n_term_out = "", c_term_out = "") {
    .Call(`_peptideUtils_oneLetterToThree`, sequences, sep_in, sep_out, n_term_out, c_term_out)
}

#' Convert from 3 letter amino acid codes to 1
#' 
#' @title Convert to 1 letter amino acid codes
#' @param sequences vector of sequences
#' @param sep_in deliminator between amino acids in input
#' @param sep_out deliminator between amino acids in output
#' @param n_term_out string to append to n terminus
#' @param c_term_out string to append to c terminus
#' @return StringVector of peptides with one letter amino acid codes
#' 
#' @examples
#' threeLetterToOne(c("Ala-Cys*-Leu-Leu-Pro", "Ala-Leu-Cys-Ala", "Ala-Gln-Sec-Ile"), sep_in = "-")
#' 
threeLetterToOne <- function(sequences, sep_in = "", sep_out = "", n_term_out = "", c_term_out = "") {
    .Call(`_peptideUtils_threeLetterToOne`, sequences, sep_in, sep_out, n_term_out, c_term_out)
}

#' Read all sequences in fasta file. Reverse matches are automatically skipped.
#'
#' @title Read fasta file.
#' 
#' @param fastaPath Path to fasta file. Be default, fasta file included in package is used.
#' @param n_entries Number of entries to read. If 0, all entires are read.
#' @return DataFrame with columns for ID and sequence.
#'
readFasta <- function(fastaPath = "", n_entries = 0L) {
    .Call(`_peptideUtils_readFasta`, fastaPath, n_entries)
}

#' Get metadata about a fasta file.
#' 
#' @title Get fasta file info.
#' 
#' @param fastaPath Path to fasta file. Be default, fasta file included in package is used.
#' @return List with slots for sequence count and vector of entry IDs contained in file.
#' 
fastaInfo <- function(fastaPath = "") {
    .Call(`_peptideUtils_fastaInfo`, fastaPath)
}

#' The function uses charge and m/z filters to remove peptides which would not be
#' observable by MS. The m/z for peptides in charge states minCharge to maxCharge
#' are calculated. If the m/z for any charge state is in between minMZ and maxMZ, the
#' sequence will be appended to peptides.
#' 
#' @title Perform a virtual protease digest of a protein.
#' 
#' @param sequences StringVector containing protein sequences. Whitespace will automatically be removed.
#' @param ids Names for the slot for each protein's peptides in output.
#' @param nMissedCleavages number of missed cleavages to allow.
#' @param cleavagePattern RegEx for protease cleavage pattern. The default is the pattern for trypsin.
#' @param mz_filter Should peptides included in output be filtered by mz?
#' @param residueAtoms Path to residueAtoms file. If blank, the default file included in the package is used. 
#' @param atomMasses Path to atomMasses file. If blank, the default file included in the package is used.
#' @param minMz Minimum m/z to allow in peptides.
#' @param maxMz Maximum m/z to allow in peptides. Set to 0 for no upper bound on m/z.
#' @param minCharge Minimum charge to consider when calculating m/z.
#' @param maxCharge Maximum charge to consider when calculating m/z.
#' @param minLen Minimum peptide length.
#' @param maxLen Maximum peptide length. Set to 0 for no upper bound on length.
#' 
#' @return A list with named elements containing vectors of each input protein's peptides.
#' 
#' @examples
#' digest(c("KLGAARKLGAGLAKVIGAGIGIGK", "KLGAARKLGAGLAKPVIGAGIGIGK"), c('a', 'b'))
#'
digest <- function(sequences, ids, nMissedCleavages = 0L, cleavagePattern = "([RK])(?=[^P])", mz_filter = TRUE, residueAtoms = "", atomMasses = "", minMz = 400, maxMz = 1800, minCharge = 1L, maxCharge = 5L, minLen = 6L, maxLen = 0L) {
    .Call(`_peptideUtils_digest`, sequences, ids, nMissedCleavages, cleavagePattern, mz_filter, residueAtoms, atomMasses, minMz, maxMz, minCharge, maxCharge, minLen, maxLen)
}

